Add stable deployment stacks swagger

kalbert312 opened the pull request:
# ARM (Control Plane) API Specification Update Pull Request 

> [!TIP]
> Overwhelmed by all this guidance? See the `Getting help` section at the bottom of this PR description.

> [!NOTE]
> As of January 2024 there is no PR assignee. This is expected. See https://aka.ms/azsdk/pr-arm-review.

## PR review workflow diagram

Please understand this diagram before proceeding. It explains how to get your PR approved & merged.

![diagram](https://github.com/Azure/azure-rest-api-specs/assets/20296335/35140fb9-639e-4e44-ac4b-38b1a20dddf9)

<details>
<summary> Click here to see the details of Step 1 </summary>

### Breaking changes review (Diagram Step 1)

If the automation determines you have breaking changes, i.e. Step 1 from the diagram applies to you,
  you must follow the [breaking changes process].  
  **IMPORTANT** This applies even if:
- The tool fails while it shouldn't, e.g. due to runtime exception, or incorrect detection of breaking changes.
- You believe there is no need for you to request breaking change approval, for any reason. 
    Such claims must be reviewed, and the process is the same.

</details>

<details> 
<summary> Click here to see the details of Step 2 </summary>

### ARM API changes review (Diagram Step 2)

- For help with Step 2 / "ARM review", see https://aka.ms/azsdk/pr-arm-review.

</details>

<details>
<summary> Click here to see the diagram footnotes</summary>

### Diagram footnotes

[1] See [ARM review queue] (for **PR merge** queues, see [2]).  
[2] [public repo merge queue], [private repo merge queue] (for **ARM review** queue, [1])  
The ARM reviewer on-call engineer visits the merge queue twice a day, so the approximate ETA for merges is 12 - 24 hours.

</details>

## Purpose of this PR

What's the purpose of this PR? Check the specific option that applies. This is **mandatory**!

  - [ ] New resource provider.
  - [x] New API version for an existing resource provider. (If API spec is not defined in TypeSpec, the PR should have been generated using [OpenAPI Hub](https://aka.ms/openapihub)).
  - [ ] Update existing version for a new feature. (This is applicable only when you are revising a private preview API version.)
  - [ ] Update existing version to fix OpenAPI spec quality issues in S360.
  - [ ] Other, please clarify:
    - _edit this with your clarification_

## Due diligence checklist

To merge this PR, you **must** go through the following checklist and confirm you understood 
and followed the instructions by checking all the boxes:

- [x] I confirm this PR is modifying Azure Resource Manager (ARM) related specifications, and not data plane related specifications.
- [x] I have reviewed following [Resource Provider guidelines](https://aka.ms/rpguidelines), including
  [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and
  [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md) (estimated time: 4 hours).  
  I understand this is required before I can proceed to the diagram Step 2, "ARM API changes review", for this PR.

## Additional information

<details>
<summary> Viewing API changes</summary>

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

</details>
<details>
<summary>Suppressing failures</summary>

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[suppressions guide](https://aka.ms/azsdk/pr-suppressions) to get approval.

</details>

## Getting help

- First, please carefully read through this PR description, from top to bottom. Please fill out the `Purpose of this PR` and `Due diligence checklist`.
- To understand what you must do next to merge this PR, see the `Next Steps to Merge` comment. It will appear within few minutes of submitting this PR and will continue to be up-to-date with current PR state.
- For guidance on fixing this PR CI check failures, see the hyperlinks provided in given failure 
  and https://aka.ms/ci-fix.
- For help with PR workflow diagram Step 2 (ARM review), see https://aka.ms/azsdk/pr-arm-review.
- If the PR CI checks appear to be stuck in `queued` state, please add a comment with contents `/azp run`.
  This should result in a new comment denoting a `PR validation pipeline` has started and the checks should be updated after few minutes.
- If the help provided by the previous points is not enough, post to https://aka.ms/azsdk/support/specreview-channel and link to this PR.

[ARM review queue]: https://aka.ms/azsdk/pr-arm-review#about-the-arm-review-queue
[public repo merge queue]: https://github.com/Azure/azure-rest-api-specs/pulls?q=is%3Aopen+is%3Apr+label%3AMergeRequested+draft%3Afalse+sort%3Acreated-asc
[private repo merge queue]: https://github.com/Azure/azure-rest-api-specs-pr/pulls?q=is%3Aopen+is%3Apr+label%3AMergeRequested+-label%3AApproved-OkToMerge+draft%3Afalse+sort%3Acreated-asc
[breaking changes process]: https://eng.ms/docs/cloud-ai-platform/azure-core/azure-core-pm-and-design/trusted-platform-pm-karimb/service-lifecycle-and-actions-team/service-lifecycle-actions-team/apex/media/launchingproductbreakingchanges#breaking-change-process-1

kalbert312 committed:
Add new stacks swagger in stable.

kalbert312 committed:
Add correlation id to stack properties.

kalbert312 committed:
Add preflight endpoints.

kalbert312 committed:
Add examples.

kalbert312 committed:
Add new stacks API version to config files.

kalbert312 committed:
Add examples for validate.

kalbert312 committed:
Fix python config.

kalbert312 committed:
Update examples to use the right api version.

kalbert312 committed:
Fix several linting problems. Need to investigate location header some more.

kalbert312 committed:
Fix subscription id formatting issue.

kalbert312 committed:
Add location response header to validate responses.

kalbert312 committed:
Unique names for deployment stack validate examples.

kalbert312 committed:
Fix duplicate example name lint warnings.

kalbert312 committed:
Set to new api version.

kalbert312 committed:
More dedupe of example names.

kalbert312 committed:
Date descending order.

kalbert312 committed:
Add deploymentStack suppressions.

kalbert312 commented:
### Swagger Validation Report

> `️❌`LintDiff: 8 Errors, 2 Warnings failed [[Detail](https://github.com/Azure/azure-rest-api-specs/pull/27529/checks?check_run_id=21078003047)]

- `OperationsAPIImplementation`: Confused on this. Is this related to the current directory structure of `Microsoft.Resources`?
- `DeleteResponseCodes`: The LRO endpoint for deleting a stack will return a 200 if the stack can be synchronously deleted. I realize this is discouraged via the guidelines but not marked with an 'X'. Is this something that can be excepted?

> `️❌`Avocado: 1 Errors, 0 Warnings failed [[Detail](https://github.com/Azure/azure-rest-api-specs/pull/27529/checks?check_run_id=21078003364)]

This looks to be related to the directory structure. Can this be excepted?

### Swagger Generation Artifacts
> `️❌`SDK Breaking Change Tracking failed [[Detail](https://github.com/Azure/azure-rest-api-specs/pull/27529/checks?check_run_id=21078004636)]

Not sure what's going on here. Is it related to a missing base commit? I recently force pushed this branch such that the first commit is the copied 2022-08-01-preview swagger without changes.

> `️❌` azure-sdk-for-python-track2 failed [[Detail](https://github.com/Azure/azure-rest-api-specs/pull/27529/checks?check_run_id=21078089114)]

Not sure what's going on here

> `️❌` azure-resource-manager-schemas failed [[Detail](https://github.com/Azure/azure-rest-api-specs/pull/27529/checks?check_run_id=21078093582)]

Not sure what's going on here

kalbert312 committed:
Fix typo.

visingla-ms commented on specification/resources/resource-manager/Microsoft.Resources/stable/2024-03-01/deploymentStacks.json:
```diff
@@ -0,0 +1,1456 @@
+            "in": "body",
+            "required": true,
+            "schema": {
+              "$ref": "#/definitions/DeploymentStack"
```
Why is DeploymentStack resource being defined here in request payload? The validate action is being performed on the same DeploymentStack resource here, right? SO, the service should already have access to resource, without needing it in the input payload.

kalbert312 replied:
It is not. This endpoint is to do adhoc validation on a stack deployment without any side effects. A user can run this via the cli and the stacks service will do validation on the stack properties as well as send off the deployment portion to the `Microsoft.Resources/deployments`'s `validate` endpoint to do preflight on the deployment which involves running preflight on each deployment's resource. Preflight on a deployment's resource will call out to the RP of that resource to do basic validation on the resource inputs as defined in the deployment details. The stack itself in this context is stored as transient data. This allows to test iterations of a deployment, changes in deployment parameters, etc, quickly. Storing it transiently also supports testing a deployment stack with new inputs when the stack already exists as a resource. After validation is successful, a user can then PUT the stack again to apply the desired changes.

kalbert312 committed:
Remove "None" for ManagedResourceReference, not a possible value.

kalbert312 committed:
Remove default value.

kalbert312 commented on specification/resources/resource-manager/Microsoft.Resources/stable/2024-03-01/deploymentStacks.json:
```diff
@@ -0,0 +1,1450 @@
+      "x-ms-parameter-location": "method",
+      "description": "Flag to indicate delete rather than detach for the resource groups."
+    },
+    "DeleteManagementGroupParameter": {
```
Leave in Swagger. Implement server-side block if necessary. If it works, implement in CLIs.

kalbert312 commented on specification/resources/resource-manager/Microsoft.Resources/stable/2024-03-01/deploymentStacks.json:
```diff
@@ -0,0 +1,1450 @@
+            "update"
+          ]
+        },
+        "parameters": {
```
Parameters: validate key vault references

kalbert312 commented on specification/resources/resource-manager/Microsoft.Resources/stable/2024-03-01/deploymentStacks.json:
```diff
@@ -0,0 +1,1450 @@
+    },
+    "DeploymentStackProperties": {
+      "type": "object",
+      "required": [
```
See if there's support for "one of required", otherwise leave as is.

kalbert312 replied:
This doesn't seem supported.

kalbert312 commented on specification/resources/resource-manager/Microsoft.Resources/stable/2024-03-01/deploymentStacks.json:
```diff
@@ -0,0 +1,1450 @@
+            "inapplicable",
+            "denyWriteAndDelete",
+            "removedBySystem",
+            "None"
```
Check what case we're returning enums in and make all consistent.

kalbert312 replied:
https://github.com/Azure/azure-rest-api-specs/pull/27529/commits/fc67dba5073dbbb612019801bbd8de0fcfd53e0e

kalbert312 committed:
Update enum casing to observed casing based on E2E tests.

kalbert312 commented on specification/resources/resource-manager/Microsoft.Resources/stable/2024-03-01/deploymentStacks.json:
```diff
@@ -0,0 +1,1450 @@
+      "properties": {
+        "location": {
+          "type": "string",
+          "x-ms-mutability": [
```
Required per https://github.com/kalbert312/azure-rest-api-specs/blob/a75560d500f595bf0805e69abbdb4410e4c958e1/documentation/openapi-authoring-automated-guidelines.md#r4002.

kalbert312 commented on specification/resources/resource-manager/Microsoft.Resources/stable/2024-03-01/deploymentStacks.json:
```diff
@@ -0,0 +1,1450 @@
+          ],
+          "description": "The location of the deployment stack. It cannot be changed after creation. It must be one of the supported Azure locations."
+        },
+        "tags": {
```
It looks to be the same: https://github.com/kalbert312/azure-rest-api-specs/blob/a75560d500f595bf0805e69abbdb4410e4c958e1/specification/resources/resource-manager/Microsoft.Resources/stable/2023-07-01/resources.json#L4730.

kalbert312 commented on specification/resources/resource-manager/Microsoft.Resources/stable/2024-03-01/deploymentStacks.json:
```diff
@@ -0,0 +1,1450 @@
+            "modelAsString": true,
+            "values": [
+              {
+                "value": "Managed",
```
Same as https://github.com/Azure/azure-rest-api-specs/pull/27529/files/dbef66e922d1529eb6ad080269275f05f71ebfbe#r1483520424

kalbert312 replied:
https://github.com/Azure/azure-rest-api-specs/pull/27529/commits/fc67dba5073dbbb612019801bbd8de0fcfd53e0e

kalbert312 commented on specification/resources/resource-manager/Microsoft.Resources/stable/2024-03-01/deploymentStacks.json:
```diff
@@ -0,0 +1,1450 @@
+              "type": "string",
+              "$ref": "#/definitions/DeploymentStacksDeleteDetachEnum"
+            },
+            "managementGroups": {
```
Same as https://github.com/Azure/azure-rest-api-specs/pull/27529/files/dbef66e922d1529eb6ad080269275f05f71ebfbe#r1483520424

kalbert312 committed:
Copy parameters type definitions from deployments service to stacks and use them for template parameters.

kalbert312 commented on specification/resources/resource-manager/Microsoft.Resources/stable/2024-03-01/deploymentStacks.json:
```diff
@@ -0,0 +1,1450 @@
+          "description": "The resourceId of the deployment resource created by the deployment stack."
+        },
+        "outputs": {
+          "type": "object",
```
I copied the types from deployments and applied them to the template parameters object in https://github.com/Azure/azure-rest-api-specs/pull/27529/commits/92a15712b839565377f7d4f06225e5d2ad8e8bd2. `outputs` is typed as a simple object in deployments though. I don't think we should update ours to be more specific than the deployments definition.

kalbert312 committed:
Apply description updates from Dante's patch.

kalbert312 committed:
Fix enum problems in x-ms-enum, default value, and examples.

kalbert312 committed:
Add lint exclusion for additional properties for deployment related properties.

rkmanda reviewed (approved):
:shipit:

kalbert312 committed:
Add DeleteManagementGroupParameter to rg/sub stack delete endpoint.

kalbert312 commented on specification/resources/resource-manager/Microsoft.Resources/stable/2024-03-01/deploymentStacks.json:
```diff
@@ -0,0 +1,1501 @@
+          },
+          {
+            "$ref": "../../../../../common-types/resource-management/v5/types.json#/parameters/ApiVersionParameter"
+          }
```
https://github.com/Azure/azure-rest-api-specs/pull/27529/commits/7007f18886f1613257a6fbd0d857582b8433e346

kalbert312 committed:
Merge branch 'main' into kylea/stacks-swagger

kalbert312 committed:
Remove "type" from nodes with "$ref".

kalbert312 committed:
Double quoute added reasons.

kalbert312 committed:
Move suppressions from policy client tag to directives block to unblock generate python track 2 SDK step due to https://github.com/Azure/azure-sdk-for-python/issues/34570

msyyc commented:
/azp run

kalbert312 committed:
Change 'locking' to 'updatingDenyAssignments'.

kalbert312 committed:
Merge branch 'main' into kylea/stacks-swagger

kalbert312 committed:
Add "bypassStackOutOfSyncError" property.

kalbert312 committed:
Try escaping the '*/' to fix the JS SDK step failing.

kazrael2119 committed:
Create sdk-suppressions.yaml

kazrael2119 committed:
Revert "Create sdk-suppressions.yaml"

This reverts commit 205a1c7f0b9421d445e41a07ffd9570b9fa02f4d.

kalbert312 committed:
Add BypassStackOutOfSyncErrorParameter to delete stack endpoints.

kalbert312 committed:
Remove x-ms-long-running-operation-options from /validate endpoints. Location is the default setting per https://github.com/Azure/autorest/blob/main/docs/extensions/readme.md#x-ms-long-running-operation-options

kalbert312 committed:
Add LRO location options explicitly and add retry-after header doc.

kalbert312 committed:
Add deployment stack validate result response and types. Adjust some description to be more consistent with other descriptions.

kalbert312 committed:
Documentation adjustments after looking at swagger preview.

kalbert312 committed:
Add 400 response for stack validate endpoints.

kalbert312 committed:
Update examples for stacks validate.

kalbert312 committed:
Merge branch 'main' into kylea/stacks-swagger

# Conflicts:
#	specification/resources/resource-manager/readme.md

kalbert312 committed:
Re add stacks stable package after merge.

kalbert312 committed:
Fix for CI step failing.

kalbert312 committed:
Add missing outer properties object in examples.

kalbert312 committed:
Fixes to validation schemas.

kalbert312 commented:
/azp run

kalbert312 commented:
I noticed in recent runs the "azure-for-python-track2" step is skipped. It was running before and I was able to download the python wheel/zip for azure-mgmt-resource. I checked the "azure-sdk-for-python" step but the artifacts listed in the check are not links like the "azure-for-python-track2" check had. I checked the AzDevOps page and didn't see the artifacts I was expecting.

Where can I download this artifact? I'm trying to test the new version of the python sdk in my local AZ CLI branch.

kalbert312 committed:
Remove non-error fields from 400 responses on /validate.

kalbert312 committed:
Add x-ms-error-response to validate 400's and add suppressions.

kalbert312 committed:
Remove status code from validate result. This is a server-side only field and will not be visible to clients.

kalbert312 committed:
After inspection of generated python code, these need to be set to false because a 400 is validation failure and should not throw an exception.

kalbert312 committed:
Update supressions.

Alancere committed:
add azure-sdk-for-go suppression

raosuhas commented:
No change in api section since last signoff so signing off again

msyyc commented:
No Python SDK breaking so the PR looks good for python.

kalbert312 commented:
There may be an issue with the ErrorResponse models for Powershell. Looking into it.

dantedallag committed:
Changed ErrorResponse to a custom object for deployment stacks.

This prevents the .NET sdk generation for powershell to not override
existing uses of a common ErrorResponse object on generation.

kalbert312 committed:
Error property should be ErrorDetail. Update validate result error ref.

kalbert312 committed:
Add "type" property to DeploymentParameter.

raosuhas commented on specification/resources/resource-manager/Microsoft.Resources/stable/2024-03-01/deploymentStacks.json:
```diff
@@ -0,0 +1,1652 @@
+        "parameters": {
+          "type": "object",
+          "description": "Deployment parameters."
+        },
```
Free form objects are not allowed in ARM .  Can you define this correctly using types. ?


If this is the same as anARM template parameter then you can use the schema of defining this as a hashtoken and have only the "value" be freeform.

raosuhas replied:
Note for future reviewers all other LintDiff errors seem to be false negatives : https://github.com/Azure/azure-rest-api-specs/pull/27529/checks?check_run_id=24038093109

kalbert312 replied:
https://github.com/Azure/azure-rest-api-specs/pull/27529/commits/1644d1e622fe02cadf13c98cda3a93cdb464d6f7

raosuhas commented:
> @raosuhas We made this change for an issue we are facing in azure-powershell with the autorest powershell generated sdk. With the deprecation of the Track 1 .NET sdk, the azure-powershell team has moved forward with a checked-in sdk solution that utilizes the new autorest powershell. We have migrated our module, Az.Resources, to utilize this. Az.Resources contains cmdlets from multiple teams, but they all share a generated sdk directory, meaning that common type generations are shared. Much of the existing checked in code relies on v1 of ErrorResponse, which gets overwritten and becomes a breaking change if used directly here. We don't have the bandwidth currently to upgrade all apis used in Az.Resources to a newer version of ErrorResponse, so this seems like the only option. Let us know if you have any questions.

I see, ok sounds good. Avoiding breaking changes are higher priority than consistency across apis. So this change is fine. You do have a few Lint errors that have come up now. I have marked it in a comment. Could you please fix that one ?

kalbert312 committed:
Add additional properties definition for validate result properties.parameters.

kalbert312 committed:
Update examples.

msyyc commented:
I am only Python SDK owner. This PR doesn't have Python SDK breaking so you don't need my approval.

kalbert312 commented:
Current breaking changes:
The current listing of added paths for `/validate` are expected. I'm not sure why the other endpoints are listed in this as they are present in the 2022-08-01-preview version. Maybe a bump of the common type version for api version and subscription triggered this?

VersioningReviewRequired:
I needed to add an escape pattern to a description to prevent the JS SDK step from crashing. I added it to both the new version and existing version.

kalbert312 committed:
Try reverting the escape sequence which may have triggered incorrect breaking changes report.

kalbert312 committed:
Add lint suppressions.

kalbert312 committed:
Revert "Try reverting the escape sequence which may have triggered incorrect breaking changes report."

This reverts commit 8f5f56c2c3e368b10b6a8e75f7716c6b348352d6.

kalbert312 commented:
Hi @ramoka178, are you able to do ARM review once more? There was an issue with PR checks that we tried to work around that required commits. No effective changes since sign off except for a couple more suppressions.

anthony-c-martin commented:
@raych1 please could you help with the `BreakingChange-Go-Sdk-Suppression` label?

kalbert312 commented:
/pr RequestMerge
